bug:  controller manager crashes on OpenShift within a few minutes of start

[cm2.log](https://github.com/kubestellar/kubestellar/files/14706234/cm2.log)
### Describe the bug

Running kubestellar v0.21.0 on OpenShift testing scenario#2 which uses WDS of type host. This cluster has many crds.

```
kubectl api-resources | wc                                                                                                                                                                                                                      ─╯
     409    1763   69102
```
 
After the KS install I deployed mcad appwrapper on the host and noticed that it was not downsyncing to the WEC. Checked the controller manager log and noticed a crash. Check the attached log details. The reason for the crash is:` panic: runtime error: index out of range [0] with length 0`

This crash happens soon after the controller manager starts.

I verified that my WECs have joined and all is well there:

```
kubectl --context imbs1 get managedclusters                                                                          

NAME  HUB ACCEPTED  MANAGED CLUSTER URLS  JOINED  AVAILABLE  AGE
wec4  true                 True   True    155m
wec3  true                 True   True    155m
```

I am running with private image built by Paolo which fixes a timeout problem in controller manager. The image is 

[quay.io/pdettori/controller-manager:1710901385](http://quay.io/pdettori/controller-manager:1710901385)


### Steps To Reproduce

Follow kubestellar install steps documented in examples for scenario#2 
[](https://docs.kubestellar.io/main/direct/examples/#scenario-2-using-the-hosting-cluster-as-wds-to-deploy-a-custom-resource)

After installing controller manager release level v0.21.0 (as instructed by KS documentation), edit controller manager deployment and replace the cm image with [quay.io/pdettori/controller-manager:1710901385](url)



### Expected Behavior

The controller manager should not crash 

### Additional Context

_No response_

Looks like I reproduced the bug. My log of the panicing controller manager:
```
E0325 21:08:51.935349       1 runtime.go:79] Observed a panic: runtime.boundsError{x:0, y:0, signed:true, code:0x0} (runtime error: index out of range [0] with length 0)
goroutine 631 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x1564620?, 0x4000e8d038})
	k8s.io/apimachinery@v0.28.2/pkg/util/runtime/runtime.go:75 +0x7c
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0x4000deb1a0?})
	k8s.io/apimachinery@v0.28.2/pkg/util/runtime/runtime.go:49 +0x78
panic({0x1564620?, 0x4000e8d038?})
	runtime/panic.go:914 +0x218
github.com/kubestellar/kubestellar/pkg/binding.pickSingleDestination(0x19297a0?)
	github.com/kubestellar/kubestellar/pkg/binding/bindingpolicy.go:403 +0xc4
github.com/kubestellar/kubestellar/pkg/binding.(*Controller).handleBindingPolicy(0x40004c2000, {0x191f348, 0x4000c0d020}, {{{0x4000ad03c0, 0x16}, {0x4000ad03d7, 0x8}, {0x40006cacc0, 0xd}}, {0x168b7f4, ...}, ...})
	github.com/kubestellar/kubestellar/pkg/binding/bindingpolicy.go:94 +0x3ac
github.com/kubestellar/kubestellar/pkg/binding.(*Controller).reconcile(0x10?, {0x191f348, 0x4000c0d020}, {{{0x4000ad03c0, 0x16}, {0x4000ad03d7, 0x8}, {0x40006cacc0, 0xd}}, {0x168b7f4, ...}, ...})
	github.com/kubestellar/kubestellar/pkg/binding/controller.go:541 +0x4c8
github.com/kubestellar/kubestellar/pkg/binding.(*Controller).processNextWorkItem.func1(0x40004c2000, {0x191f348?, 0x4000c0d020?}, {{0x1921688?, 0x4000253f80?}, 0x40004d1760?}, {0x14dbe60?, 0x4000deb1a0?})
	github.com/kubestellar/kubestellar/pkg/binding/controller.go:512 +0x1c8
github.com/kubestellar/kubestellar/pkg/binding.(*Controller).processNextWorkItem(0x40004c2000, {0x191f348, 0x4000c0d020})
	github.com/kubestellar/kubestellar/pkg/binding/controller.go:522 +0x120
github.com/kubestellar/kubestellar/pkg/binding.(*Controller).runWorker(...)
	github.com/kubestellar/kubestellar/pkg/binding/controller.go:473
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1()
	k8s.io/apimachinery@v0.28.2/pkg/util/wait/backoff.go:259 +0x30
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x0?)
	k8s.io/apimachinery@v0.28.2/pkg/util/wait/backoff.go:226 +0x40
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x4000c046c8?, {0x19060c0, 0x4000db8e10}, 0x1, 0x0)
	k8s.io/apimachinery@v0.28.2/pkg/util/wait/backoff.go:227 +0x90
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x1037b50?, 0x3b9aca00, 0x0, 0xb0?, 0x4000642ae0?)
	k8s.io/apimachinery@v0.28.2/pkg/util/wait/backoff.go:204 +0x80
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext({0x191f348, 0x4000c0d020}, 0x4000c0ea50, 0x47ce7c?, 0x4000596300?, 0x0?)
	k8s.io/apimachinery@v0.28.2/pkg/util/wait/backoff.go:259 +0x80
k8s.io/apimachinery/pkg/util/wait.UntilWithContext({0x191f348?, 0x4000c0d020?}, 0x4000253d40?, 0x4000c04708?)
	k8s.io/apimachinery@v0.28.2/pkg/util/wait/backoff.go:170 +0x2c
created by github.com/kubestellar/kubestellar/pkg/binding.(*Controller).run in goroutine 105
	github.com/kubestellar/kubestellar/pkg/binding/controller.go:354 +0x900
panic: runtime error: index out of range [0] with length 0 [recovered]
	panic: runtime error: index out of range [0] with length 0
```
Comparing to the panic in cm2.log:
```
E0321 18:03:25.792358       1 runtime.go:79] Observed a panic: runtime.boundsError{x:0, y:0, signed:true, code:0x0} (runtime error: index out of range [0] with length 0)
goroutine 2332 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x1b11240?, 0xc004caccd8})
	k8s.io/apimachinery@v0.28.2/pkg/util/runtime/runtime.go:75 +0x99
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc000ff8180?})
	k8s.io/apimachinery@v0.28.2/pkg/util/runtime/runtime.go:49 +0x75
panic({0x1b11240, 0xc004caccd8})
	runtime/panic.go:884 +0x213
github.com/kubestellar/kubestellar/pkg/binding.pickSingleDestination(0x1ecd2c0?)
	github.com/kubestellar/kubestellar/pkg/binding/bindingpolicy.go:398 +0xc5
github.com/kubestellar/kubestellar/pkg/binding.(*Controller).handleBindingPolicy(0xc000c560d0, {0x1ec3790, 0xc00468d470}, {{{0xc000a0e620, 0x16}, {0xc000a0e637, 0x8}, {0xc000c114c0, 0xd}}, {0xc000e328b0, ...}, ...})
	github.com/kubestellar/kubestellar/pkg/binding/bindingpolicy.go:93 +0x4bf
github.com/kubestellar/kubestellar/pkg/binding.(*Controller).reconcile(0xc000ff81c0?, {0x1ec3790, 0xc00468d470}, {{{0xc000a0e620, 0x16}, {0xc000a0e637, 0x8}, {0xc000c114c0, 0xd}}, {0xc000e328b0, ...}, ...})
	github.com/kubestellar/kubestellar/pkg/binding/controller.go:536 +0x59c
github.com/kubestellar/kubestellar/pkg/binding.(*Controller).processNextWorkItem.func1(0xc000c560d0, {0x1ec3790?, 0xc00468d470?}, {{0x1ec62f0?, 0xc000c9cf40?}, 0xc004690980?}, {0x1a900e0?, 0xc000ff8180?})
	github.com/kubestellar/kubestellar/pkg/binding/controller.go:507 +0x2a5
github.com/kubestellar/kubestellar/pkg/binding.(*Controller).processNextWorkItem(0xc000c560d0, {0x1ec3790, 0xc00468d470})
	github.com/kubestellar/kubestellar/pkg/binding/controller.go:517 +0x154
github.com/kubestellar/kubestellar/pkg/binding.(*Controller).runWorker(...)
	github.com/kubestellar/kubestellar/pkg/binding/controller.go:468
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1()
	k8s.io/apimachinery@v0.28.2/pkg/util/wait/backoff.go:259 +0x25
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x30?)
	k8s.io/apimachinery@v0.28.2/pkg/util/wait/backoff.go:226 +0x3e
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x0?, {0x1eac240, 0xc00468d4a0}, 0x1, 0x0)
	k8s.io/apimachinery@v0.28.2/pkg/util/wait/backoff.go:227 +0xb6
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x0?, 0x3b9aca00, 0x0, 0xc?, 0xc000c386e0?)
	k8s.io/apimachinery@v0.28.2/pkg/util/wait/backoff.go:204 +0x89
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext({0x1ec3790, 0xc00468d470}, 0xc0046c9fa0, 0x4?, 0x0?, 0x0?)
	k8s.io/apimachinery@v0.28.2/pkg/util/wait/backoff.go:259 +0x99
k8s.io/apimachinery/pkg/util/wait.UntilWithContext({0x1ec3790?, 0xc00468d470?}, 0x0?, 0x0?)
	k8s.io/apimachinery@v0.28.2/pkg/util/wait/backoff.go:170 +0x2b
created by github.com/kubestellar/kubestellar/pkg/binding.(*Controller).run
	github.com/kubestellar/kubestellar/pkg/binding/controller.go:349 +0xc1f
panic: runtime error: index out of range [0] with length 0 [recovered]
	panic: runtime error: index out of range [0] with length 0
```
Looks like the two call stacks are identical.

--- a/pkg/binding/bindingpolicy.go
+++ b/pkg/binding/bindingpolicy.go
@@ -403,7 +403,8 @@ func SliceContains[Elt comparable](slice []Elt, seek Elt) bool {
 	return false
 }
 
-// sort by name and pick first cluster so that the choice is deterministic based on names
+// pickSingleDestination sorts clusters by name and picks first cluster so that the choice is deterministic based on names
+// pickSingleDestination expects a non-empty clusterSet
 func pickSingleDestination(clusterSet sets.Set[string]) sets.Set[string] {
 	return sets.New(sets.List(clusterSet)[0])
 }

--- a/pkg/binding/bindingpolicy.go
+++ b/pkg/binding/bindingpolicy.go
@@ -86,12 +86,17 @@ func (c *Controller) handleBindingPolicy(ctx context.Context, objIdentifier util
 		if err != nil {
 			return fmt.Errorf("failed to ocm.FindClustersBySelectors: %w", err)
 		}
+		if len(clusterSet) == 0 {
+			logger.Info("No clusters are selected by BindingPolicy", "name", bindingPolicy.Name)
+		}
 
 		if bindingPolicy.Spec.WantSingletonReportedState {
 			// if the bindingpolicy requires a singleton status, then we should only
 			// have one destination
 			// TODO: this should be removed once we have proper enforcement or error reporting for this
-			clusterSet = pickSingleDestination(clusterSet)
+			if len(clusterSet) > 1 {
+				clusterSet = pickSingleDestination(clusterSet)
+			}
 		}
 
 		// set destinations and enqueue binding for syncing

--- a/test/e2e/ginkgo/multiple_cluster_deployment_test.go
+++ b/test/e2e/ginkgo/multiple_cluster_deployment_test.go
@@ -214,6 +214,29 @@ var _ = ginkgo.Describe("end to end testing", func() {
 	})
 
 	ginkgo.Context("singleton status testing", func() {
+		ginkgo.It("no singleton status when a singleton bindingpolicy/deployment is created but no matching WEC(s)", func() {
+			util.DeleteDeployment(ctx, wds, ns, "nginx")
+			util.CreateDeployment(ctx, wds, ns, "nginx-singleton",
+				map[string]string{
+					"app.kubernetes.io/name": "nginx-singleton",
+				})
+			util.CreateBindingPolicy(ctx, ksWds, "nginx-singleton",
+				[]metav1.LabelSelector{
+					{MatchLabels: map[string]string{"name": "CelestialNexus"}},
+				},
+				[]ksapi.DownsyncObjectTest{
+					{ObjectSelectors: []metav1.LabelSelector{
+						{MatchLabels: map[string]string{"app.kubernetes.io/name": "nginx-singleton"}},
+					}}})
+			patch := []byte(`{"spec":{"wantSingletonReportedState": true}}`)
+			_, err := ksWds.ControlV1alpha1().BindingPolicies().Patch(
+				ctx, "nginx-singleton", types.MergePatchType, patch, metav1.PatchOptions{})
+			gomega.ExpectWithOffset(1, err).NotTo(gomega.HaveOccurred())
+			util.ValidateNumDeployments(ctx, wec1, ns, 0)
+			util.ValidateNumDeployments(ctx, wec2, ns, 0)
+			util.ValidateSingletonStatusZeroValue(ctx, wds, ns, "nginx-singleton")
+		})
+
 		ginkgo.It("sets singleton status when a singleton bindingpolicy/deployment is created", func() {
 			util.DeleteDeployment(ctx, wds, ns, "nginx") // we don't have to delete nginx
 			util.CreateDeployment(ctx, wds, ns, "nginx-singleton",

--- a/test/e2e/ginkgo/multiple_cluster_deployment_test.go
+++ b/test/e2e/ginkgo/multiple_cluster_deployment_test.go
@@ -214,29 +214,6 @@ var _ = ginkgo.Describe("end to end testing", func() {
 	})
 
 	ginkgo.Context("singleton status testing", func() {
-		ginkgo.It("no singleton status when a singleton bindingpolicy/deployment is created but no matching WEC(s)", func() {
-			util.DeleteDeployment(ctx, wds, ns, "nginx")
-			util.CreateDeployment(ctx, wds, ns, "nginx-singleton",
-				map[string]string{
-					"app.kubernetes.io/name": "nginx-singleton",
-				})
-			util.CreateBindingPolicy(ctx, ksWds, "nginx-singleton",
-				[]metav1.LabelSelector{
-					{MatchLabels: map[string]string{"name": "CelestialNexus"}},
-				},
-				[]ksapi.DownsyncObjectTest{
-					{ObjectSelectors: []metav1.LabelSelector{
-						{MatchLabels: map[string]string{"app.kubernetes.io/name": "nginx-singleton"}},
-					}}})
-			patch := []byte(`{"spec":{"wantSingletonReportedState": true}}`)
-			_, err := ksWds.ControlV1alpha1().BindingPolicies().Patch(
-				ctx, "nginx-singleton", types.MergePatchType, patch, metav1.PatchOptions{})
-			gomega.ExpectWithOffset(1, err).NotTo(gomega.HaveOccurred())
-			util.ValidateNumDeployments(ctx, wec1, ns, 0)
-			util.ValidateNumDeployments(ctx, wec2, ns, 0)
-			util.ValidateSingletonStatusZeroValue(ctx, wds, ns, "nginx-singleton")
-		})
-
 		ginkgo.It("sets singleton status when a singleton bindingpolicy/deployment is created", func() {
 			util.DeleteDeployment(ctx, wds, ns, "nginx") // we don't have to delete nginx
 			util.CreateDeployment(ctx, wds, ns, "nginx-singleton",
@@ -258,6 +235,39 @@ var _ = ginkgo.Describe("end to end testing", func() {
 			util.ValidateNumDeployments(ctx, wec1, ns, 1)
 			util.ValidateNumDeployments(ctx, wec2, ns, 0)
 			util.ValidateSingletonStatus(ctx, wds, ns, "nginx-singleton")
+		})
+
+		// This ginkgo subject node is currently written in an 'incorrect' way, in the sense that
+		// the Deployment's status always begins with zero value no matter there are matching WEC(s) or not, so
+		// this node doesn't test any behaviors of KubeStellar.
+		// The correct way to test is to merge this node with the previous node, to show that the Deployment's status:
+		// (1) is updated (with non-zero value) when there are matching WEC(s); then
+		// (2) is deleted (ends up with zero value) when there are no matching WEC(s) any more, due to changes of the bindingpolicy's ClusterSelectors or WEC(s)'s labels.
+		//
+		// But today there is a bug which prevents such a merged node from succeeding.
+		// We will complete the tests here by merging the two nodes after fixing that bug.
+		// We track this by https://github.com/kubestellar/kubestellar/issues/2020#issuecomment-2037539110
+		ginkgo.It("zeros singleton status when there is a singleton bindingpolicy/deployment but WEC(s) are not selected any more", func() {
+			util.DeleteDeployment(ctx, wds, ns, "nginx")
+			util.CreateDeployment(ctx, wds, ns, "nginx-singleton",
+				map[string]string{
+					"app.kubernetes.io/name": "nginx-singleton",
+				})
+			util.CreateBindingPolicy(ctx, ksWds, "nginx-singleton",
+				[]metav1.LabelSelector{
+					{MatchLabels: map[string]string{"name": "CelestialNexus"}},
+				},
+				[]ksapi.DownsyncObjectTest{
+					{ObjectSelectors: []metav1.LabelSelector{
+						{MatchLabels: map[string]string{"app.kubernetes.io/name": "nginx-singleton"}},
+					}}})
+			patch := []byte(`{"spec":{"wantSingletonReportedState": true}}`)
+			_, err := ksWds.ControlV1alpha1().BindingPolicies().Patch(
+				ctx, "nginx-singleton", types.MergePatchType, patch, metav1.PatchOptions{})
+			gomega.ExpectWithOffset(1, err).NotTo(gomega.HaveOccurred())
+			util.ValidateNumDeployments(ctx, wec1, ns, 0)
+			util.ValidateNumDeployments(ctx, wec2, ns, 0)
+			util.ValidateSingletonStatusZeroValue(ctx, wds, ns, "nginx-singleton")
 		})
 	})
 

--- a/test/util/util.go
+++ b/test/util/util.go
@@ -310,6 +310,14 @@ func ValidateNumJobs(ctx context.Context, wec *kubernetes.Clientset, ns string, 
 	}, timeout).Should(gomega.Equal(num))
 }
 
+func ValidateSingletonStatusZeroValue(ctx context.Context, wds *kubernetes.Clientset, ns string, name string) {
+	gomega.Eventually(func() []appsv1.DeploymentCondition {
+		deployment, err := wds.AppsV1().Deployments(ns).Get(ctx, name, metav1.GetOptions{})
+		gomega.ExpectWithOffset(1, err).NotTo(gomega.HaveOccurred())
+		return deployment.Status.Conditions
+	}, timeout).Should(gomega.BeNil())
+}
+
 func ValidateSingletonStatus(ctx context.Context, wds *kubernetes.Clientset, ns string, name string) {
 	gomega.Eventually(func() int {
 		deployment, err := wds.AppsV1().Deployments(ns).Get(ctx, name, metav1.GetOptions{})

